### DISCUSSION TITLE
When will the 7.0 branch be released?

### DISCUSSION THREAD

**Discussion by Daliji** [vladimir-bukhtoyarov/bucket4j#161] [General]
I saw version 5.0.0-beta.01-09-2020 in maven repository, in order to support redis based distributed token bucket,can I use this version in my project for a long support?

**Comment by vladimir-bukhtoyarov** [comment]
Hello @Daliji,

I plan to officially release ```7.0```at the end of May.  Next weekend I can release the version ```7.0.0.beta``` which will be suitable for long support(but without accurate javadocs and readme pages). But before releasing ```7.0.0.beta``` I want to ask you how do you plan to use Redis support? Do you mean to use this [RMapBasedRedissonBackend](https://github.com/vladimir-bukhtoyarov/bucket4j/blob/7.0/bucket4j-examples/src/test/java/example/distributed/generic/redisson/rmap_cas/RMapBasedRedissonBackend.java)? I just want to notice that this issue #36 is not a subject for ```7.0```.  RMapBasedRedissonBackend by the link above is just an example that is based on CAS and from a performance perspective, it can not be as fast as the solution on Lua Stored procedures which will be implemented in the scope of #36

**Reply by Daliji** [reply]
Hello @vladimir-bukhtoyarov  ,

Yes I want to use RMapBasedRedissonBackend for a distributed rate limiter. My boss do not want to use Hazelcast or other components in our projects. I am looking for a mature library to replace Redis-Lua solution.

And I'm also very much looking forward to supporting Redis' distributed rate limiter release.

**Reply by ph4r05** [reply]
Distributed Redis-based RateLimiter would be great for AWS environment as it already has managed Redis clusters. We are looking for this as well.

**Reply by vladimir-bukhtoyarov** [reply]
@ph4r05 as was described in the comments above, I am going to release the beta version(without javadocs) at this weekend.

**Comment by vladimir-bukhtoyarov** [comment]
I have released 7.0.0-beta-2020-04-18 just now. @Daliji you need to use this class https://github.com/vladimir-bukhtoyarov/bucket4j/blob/7.0/bucket4j-redis/src/main/java/io/github/bucket4j/redis/redisson/cas/RMapBasedRedissonProxyManager.java

It is interesting to see benchmarking results versus your custom solution. I expect that RMapBasedRedissonProxyManager should be less effective than your solution because RMapBasedRedissonProxyManager is based on CAS operations instead of Lua stored procedures, and CAS should behave less effective in case of high contention on the same key.

**Reply by bstern** [reply]
For a newbie investigating Bucket4j and Redis integration, how/where does RMapBasedRedissonProxyManager.java fit in? Does it support named references? Thanks.

**Reply by vladimir-bukhtoyarov** [reply]
@bstern it is better to ask [Redisson developers](https://github.com/redisson/redisson) directly, I just use Redisson [RMap API](https://www.javadoc.io/doc/org.redisson/redisson/3.4.1/org/redisson/api/RMap.html) without any inside about internal implementation.

**Reply by bstern** [reply]
Thanks, I was more asking how or where to reference this class in Bucket4j so Redis is used. I apologize since I'm at the very beginning of learning this project.

**Reply by vladimir-bukhtoyarov** [reply]
All usage of Redisson is localized to single class - [RMapBasedRedissonProxyManager](https://github.com/vladimir-bukhtoyarov/bucket4j/blob/7.0/bucket4j-redis/src/main/java/io/github/bucket4j/redis/redisson/cas/RMapBasedRedissonProxyManager.java).

**Comment by apascenko** [comment]
Hi @vladimir-bukhtoyarov, I'm evaluating the possibility of using `RMapBasedRedissonProxyManager` for a distributed rate limiter. I have a question: do the buckets expire after some time? I.e., if I have a significant number of users that run a large volume of requests (due to some automated processing) only occasionally, would the buckets for those users stay in Redis forever or would they eventually get cleaned up?

**Reply by vladimir-bukhtoyarov** [reply]
Hello @apascenko 

Bucket4j does not specify write expiry. I hope it is possible to configure expiration via RMap interface.

**Comment by pelepelin** [comment]
Hello @vladimir-bukhtoyarov !

I have a question, is there a reason why RMap is used?

As I see, RMap basic implementations RedissonMap and RedissonMapCache map all bucket4j buckets into a single Redis hashmap. Consequently, RedissonMap does not provide per-bucket expiration which is unsuitable for large bucket sets. RedissonMapCache uses quite complex script and application-side worker for bucket expiration, which is quite unnatural. Also since all buckets are stored under a single Redis hashmap, Redis Cluster can't be used naturally (RClusteredMap uses Redisson custom sharding implementation).

I would expect each bucket4j bucket to map naturally to a single Redis key-value pair (probably, using RBucket) which would naturally support Redis Cluster and expiration.

**Reply by vladimir-bukhtoyarov** [reply]
Hello @pelepelin 

Thanks for pointing it out, I will investigate the mentioned problems a bit later.

**Reply by vladimir-bukhtoyarov** [reply]
@pelepelin looks that RBucket is unsutable too because its API https://www.javadoc.io/doc/org.redisson/redisson/latest/org/redisson/api/RBucket.html does not provide compare&swap with TTL. So it is necessary to implement something handmade on top of Redisson command executor.

**Reply by vladimir-bukhtoyarov** [reply]
I have rewritten Redis support via low-level Reddison-API https://github.com/vladimir-bukhtoyarov/bucket4j/commit/b2d1bbb59e22f63595861e3be6e2e629ca80f595

**Reply by alxvsk** [reply]
Hello @vladimir-bukhtoyarov 
are you planning to include Redis support in the upcoming 7.0 release?

**Reply by vladimir-bukhtoyarov** [reply]
@alxvsk 

I have rewritten the wrong implementation, so yes, it will be included.
